File: src/main/java/nl/chimpgamer/networkmanager/extensions/discordbot/utils/tasks/GuildJoinCheckTask.java
package nl.chimpgamer.networkmanager.extensions.discordbot.utils.tasks;

import lombok.RequiredArgsConstructor;
import net.dv8tion.jda.core.entities.Member;

import net.dv8tion.jda.core.entities.Role;
import nl.chimpgamer.networkmanager.api.cache.modules.CachedPlayers;
import nl.chimpgamer.networkmanager.extensions.discordbot.DiscordBot;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.Utils;
import nl.chimpgamer.networkmanager.api.models.player.Player;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Optional;
import java.util.UUID;

@RequiredArgsConstructor
public class GuildJoinCheckTask implements Runnable {
    private final DiscordBot discordBot;
    private final Member member;

    @Override
    public void run() {
        final CachedPlayers cachedPlayers = this.getDiscordBot().getNetworkManager().getCacheManager().getCachedPlayers();

        try (Connection connection = this.getDiscordBot().getNetworkManager().getMySQL().getConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT DiscordID, UUID FROM nm_discordusers WHERE DiscordID=?;")) {
            ps.setString(1, this.getMember().getUser().getId());
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    Optional<Player> opPlayer = cachedPlayers.getPlayerSafe(UUID.fromString(rs.getString("UUID")));
                    if (!opPlayer.isPresent()) {
                        return;
                    }
                    Player player = opPlayer.get();

                    if (this.getDiscordBot().getConfigManager().isVerifyAddRole()) {
                        String verifyRoleName = this.getDiscordBot().getConfigManager().getVerifyRole();
                        if (!verifyRoleName.isEmpty()) {
                            Role verifiedRole = Utils.getRoleByName(verifyRoleName);
                            if (verifiedRole != null) {
                                this.getDiscordBot().getGuild().getController().addSingleRoleToMember(member, verifiedRole).queue();
                            } else {
                                this.getDiscordBot().warn("The verified role '" + verifyRoleName + "' does not seem to exist!");
                            }
                        } else {
                            this.getDiscordBot().warn("The verified role '" + verifyRoleName + "' does not seem to exist!");
                        }
                    }


                    if (this.getDiscordBot().getConfigManager().isSyncUserName()) {
                        Utils.setNickName(this.getMember(), player.getName());
                    }

                    if (this.getDiscordBot().getConfigManager().isSyncRanks()) {
                        Utils.syncRanks(player);
                    }
                }
            }
        } catch (SQLException ex) {
            ex.printStackTrace();
        }
    }

    private Member getMember() {
        return member;
    }

    private DiscordBot getDiscordBot() {
        return discordBot;
    }
}

File: shared/src/main/kotlin/nl/chimpgamer/networkmanager/extensions/discordbot/shared/tasks/GuildJoinCheckTask.kt
package nl.chimpgamer.networkmanager.extensions.discordbot.shared.tasks

import net.dv8tion.jda.api.entities.Member
import nl.chimpgamer.networkmanager.api.utils.Placeholders
import nl.chimpgamer.networkmanager.extensions.discordbot.shared.DiscordBot
import nl.chimpgamer.networkmanager.extensions.discordbot.shared.configurations.DCMessage
import nl.chimpgamer.networkmanager.extensions.discordbot.shared.configurations.Setting
import java.sql.SQLException
import java.util.UUID

class GuildJoinCheckTask(private val discordBot: DiscordBot, private val member: Member) : Runnable {

    override fun run() {
        val cachedPlayers = discordBot.networkManager.cacheManager.cachedPlayers
        discordBot.networkManager.storage.connection.use { connection ->
            try {
                connection.prepareStatement("SELECT DiscordID, UUID FROM nm_discordusers WHERE DiscordID=?;").use { ps ->
                    ps.setString(1, member.user.id)
                    ps.executeQuery().use { rs ->
                        if (rs.next()) {
                            val player = cachedPlayers.getPlayer(UUID.fromString(rs.getString("UUID"))) ?: return

                            val syncUsername = discordBot.settings.getBoolean(Setting.DISCORD_SYNC_USERNAME_ENABLED)
                            val addVerifiedRole = discordBot.settings.getBoolean(Setting.DISCORD_REGISTER_ADD_ROLE_ENABLED) && discordBot.discordManager.verifiedRole != null

                            if (syncUsername && addVerifiedRole) {
                                val nickname = Placeholders.setPlaceholders(player, discordBot.settings.getString(Setting.DISCORD_SYNC_USERNAME_FORMAT))
                                discordBot.discordManager.setNickNameAndAddRole(member, nickname, discordBot.discordManager.verifiedRole!!)
                            } else {
                                if (syncUsername) {
                                    val nickname = Placeholders.setPlaceholders(player, discordBot.settings.getString(Setting.DISCORD_SYNC_USERNAME_FORMAT))
                                    discordBot.discordManager.setNickName(member, nickname)
                                }
                                if (addVerifiedRole) {
                                    discordBot.discordManager.verifiedRole?.let { discordBot.discordManager.addRoleToMember(member, it) }
                                }
                            }

                            if (discordBot.settings.getBoolean(Setting.DISCORD_SYNC_RANKS_ENABLED)) {
                                discordBot.scheduler.runSync(SyncRanksTask(discordBot, player))
                            }
                        } else {
                            val welcomeMessage = discordBot.messages.getString(DCMessage.EVENT_JOIN)
                            if (welcomeMessage.isEmpty()) {
                                return
                            }
                            member.user.openPrivateChannel().queue { it.sendMessage(welcomeMessage.replace("%mention%", member.user.asMention)).queue() }
                        }
                    }
                }
            } catch (ex: SQLException) {
                ex.printStackTrace()
            }
        }
    }
}